engine/crates/federated-graph/src/from_sdl.rs

```diff
@@ -1014,4 +1053,180 @@ fn test_from_sdl() {
           reviews: [Review] @join__field(graph: REVIEWS)
         }
     "#).unwrap();
+
+    let schema = schema.into_latest();
+    let query_object = &schema[schema.root_operation_types.query];
+
+    for field_name in ["__type", "__schema"] {
+        let field_name = schema.strings.iter().position(|s| s == field_name).unwrap();
+        assert!(schema[query_object.fields.clone()]
+            .iter()
+            .any(|f| f.name.0 == field_name));
+    }
+}
+
+#[cfg(test)]
+#[test]
+fn test_from_sdl_with_empty_query_root() {
+    // https://github.com/the-guild-org/gateways-benchmark/blob/main/federation-v1/gateways/apollo-router/supergraph.graphql
+    let schema = super::from_sdl(
+        r#"
+        schema
+          @link(url: "https://specs.apollo.dev/link/v1.0")
+          @link(url: "https://specs.apollo.dev/join/v0.3", for: EXECUTION)
+        {
+          query: Query
+        }
+
+        directive @join__enumValue(graph: join__Graph!) repeatable on ENUM_VALUE
+
+        directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet, type: String, external: Boolean, override: String, usedOverridden: Boolean) repeatable on FIELD_DEFINITION | INPUT_FIELD_DEFINITION
+
+        directive @join__graph(name: String!, url: String!) on ENUM_VALUE
+
+        directive @join__implements(graph: join__Graph!, interface: String!) repeatable on OBJECT | INTERFACE
+
+        directive @join__type(graph: join__Graph!, key: join__FieldSet, extension: Boolean! = false, resolvable: Boolean! = true, isInterfaceObject: Boolean! = false) repeatable on OBJECT | INTERFACE | UNION | ENUM | INPUT_OBJECT | SCALAR
+
+        directive @join__unionMember(graph: join__Graph!, member: String!) repeatable on UNION
+
+        directive @link(url: String, as: String, for: link__Purpose, import: [link__Import]) repeatable on SCHEMA
+
+        scalar join__FieldSet
+
+        enum join__Graph {
+          ACCOUNTS @join__graph(name: "accounts", url: "http://accounts:4001/graphql")
+          INVENTORY @join__graph(name: "inventory", url: "http://inventory:4002/graphql")
+          PRODUCTS @join__graph(name: "products", url: "http://products:4003/graphql")
+          REVIEWS @join__graph(name: "reviews", url: "http://reviews:4004/graphql")
+        }
+
+        scalar link__Import
+
+        enum link__Purpose {
+          """
+          `SECURITY` features provide metadata necessary to securely resolve fields.
+          """
+          SECURITY
+
+          """
+          `EXECUTION` features provide metadata necessary for operation execution.
+          """
+          EXECUTION
+        }
+
+        type Query
```

Comment:
Should we actively be trying to support this?  It's not a valid schema, so doesn't seem like something we should put effort into...

Reply:
No indeed, I was just writing (and pushed) a test to make it an error.